[22744] Fast DDS Spy basic tutorial (#286)
eProsima/vulcanexus: docs/rst/tutorials/tools/fastddsspy/fastddsspy.rst

.. _tutorials_tools_fastddsspy:

Getting Started with Fast DDS Spy
==================================

Background
----------

Vulcanexus integrates `eProsima Fast DDS Spy <https://fast-dds-spy.readthedocs.io/en/latest/index.html>`_,

> an interactive command-line tool that allows inspecting DDS entities such as DomainParticipants, DataWriters,

Reviewer: ```suggestion
an interactive command-line tool that allows you to inspect DDS entities such as DomainParticipants, DataWriters,
```

    for alternative installation options.

Scenario Overview
-----------------

In this tutorial, *ShapesDemo* publishes data on a DDS topic, and *Fast DDS Spy* is used to inspect what is happening
in the DDS network.

.. code-block:: text

    ShapesDemo publisher
         |
         v
    DDS network
         |
         v
    Fast DDS Spy

The goal is to observe an existing DDS system and understand what entities and data are visible,
not to `create a publisher or subscriber from scratch <https://fast-dds.docs.eprosima.com/en/latest/fastdds/getting_started/simple_app/simple_app.html>`_.

Step 1: Start ShapesDemo
------------------------

Open *ShapesDemo* in a terminal:

.. code-block:: bash

    ShapesDemo

You should see the *ShapesDemo* application open.

.. figure:: /rst/figures/tutorials/tools/fastddsspy/ShapesDemo.png
    :align: center

Configure the participant
^^^^^^^^^^^^^^^^^^^^^^^^^

In the toolbar at the top, open the ``Options`` menu and select ``Participant Configuration``.
This window lets you choose the transport mode, domain ID, and other advanced options.

> In this tutorial, set the Domain ID to ``10``.

Reviewer: ```suggestion
This window lets you choose the transport mode, Domain ID, and other advanced options.
In this tutorial, set the Domain ID to ``10``.
```

    Commands available and the information they show:
        help                                      : this help.
        version                                   : tool version.
        quit                                      : exit interactive CLI and close program.
        participants                              : DomainParticipants discovered in the network.
        participants verbose                      : verbose information about DomainParticipants discovered in the network.
        participants <Guid>                       : verbose information related with a specific DomainParticipant.
        writers                                   : DataWriters discovered in the network.
        writers verbose                           : verbose information about DataWriters discovered in the network.
        writers <Guid>                            : verbose information related with a specific DataWriter.
        reader                                    : DataReaders discovered in the network.
        reader verbose                            : verbose information about DataReaders discovered in the network.

>         reader <Guid>                             : verbose information related with a specific DataReader.

Reviewer: We should create a Fast-DDS-Spy PR changing the `reader` output in the help command

      datawriters:
        - Square [ShapeType] (1)
      datareaders:
        - Square [ShapeType] (1)

Step 5: Discover DataWriters
------------------------------

Run:

.. code-block:: bash

    writers

You should see one DataWriter associated with the ``Square`` topic.

.. code-block:: text

    >> writers
    - guid: 01.0f.91.a5.1f.00.1c.08.00.00.01.00|0.0.1.2
      participant: Fast DDS ShapesDemo Participant
      topic: Square [ShapeType]

A DataWriter is a DDS entity that publishes data. Seeing a writer associated with ``Square`` confirms that *Fast DDS Spy* can discover the publisher side of the DDS communication.

Verbose output
^^^^^^^^^^^^^^

Running the command with the verbose option shows more information about the topic and its QoS.

.. code-block:: bash

    writers verbose

.. code-block:: text

    >> writers verbose
    - guid: 01.0f.91.a5.1f.00.1c.08.00.00.01.00|0.0.1.2
      participant: Fast DDS ShapesDemo Participant
      topic:
        name: Square
        type: ShapeType
        partitions: ""
      qos:
        durability: volatile
        reliability: reliable

Step 6: Discover DataReaders
------------------------------

Run:

.. code-block:: bash

>     readers

Reviewer: The command `readers` is the same as `reader` but is not implicit in this documentation PR. So for convenience use `reader`.

Reviewer: Revert it to have `readers`

This PR update the help command to have it as `readers` instead of `reader`
https://github.com/eProsima/Fast-DDS-spy/pull/151

You should see one DataReader associated with the ``Square`` topic.

.. code-block:: text

>     >> readers

Reviewer: Same `readers`= `reader` command. Use `reader`

Reviewer: Revert it to have `readers`

This PR update the help command to have it as `readers` instead of `reader`
https://github.com/eProsima/Fast-DDS-spy/pull/151

    - guid: 01.0f.91.a5.1f.00.1c.08.00.00.01.00|0.0.2.7
      participant: Fast DDS ShapesDemo Participant
      topic: Square [ShapeType]

A DataReader is a DDS entity that subscribes to data. Seeing a reader associated with the ``Square`` topic confirms that *Fast DDS Spy* can discover the subscriber side of the communication.

Verbose output
^^^^^^^^^^^^^^

Running this command with the verbose option shows more detailed information about the reader's topic and QoS.

.. code-block:: bash

>     readers verbose

Reviewer: Same `readers`= `reader` command. Use `reader`

Reviewer: Revert it to have `readers`

This PR update the help command to have it as `readers` instead of `reader`
https://github.com/eProsima/Fast-DDS-spy/pull/151

.. code-block:: text

>     >> readers verbose

Reviewer: Same `readers`= `reader` command. Use `reader`

Reviewer: Revert it to have `readers`

This PR update the help command to have it as `readers` instead of `reader`
https://github.com/eProsima/Fast-DDS-spy/pull/151

    - guid: 01.0f.91.a5.1f.00.1c.08.00.00.01.00|0.0.2.7
      participant: Fast DDS ShapesDemo Participant
      topic:
        name: Square
        type: ShapeType
        partitions: ""
      qos:
        durability: volatile
        reliability: best-effort

.. raw:: html

    <video width=100% height=auto autoplay loop controls>
        <source src="../../../../_static/EntitiesProperties.mp4">
        Your browser does not support the video tag.
    </video>
    <br></br>

Step 7: List DDS Topics
------------------------

Run:

.. code-block:: bash

    topics

You should see the active topics discovered in the DDS network.

.. code-block:: text

    >> topics
    - topic: Square (ShapeType) (1|1) [13.334426 Hz]

*Fast DDS Spy* has discovered the topic names used by the DDS network. Listing topics is usually the first command to run when you want to understand what data is available.

Step 8: Inspect a Topic in Detail
-----------------------------------

Run:

.. code-block:: bash

    topics Square v

> 

Reviewer: I think it is better to have the commands flow like this command1, output1, command2, output2
instead of command1, command2, output1, outpu2

      rate: 13.3355 Hz
      dynamic_type_discovered: true

Step 9: Display the Topic IDL
-------------------------------

Run:

.. code-block:: bash

    topics Square idl

*Fast DDS Spy* will display the IDL/type definition associated with the ``Square`` topic.

.. code-block:: text

    >> topics Square idl

    @extensibility(APPENDABLE)
    struct ShapeType
    {
        @key string color;
        long x;
        long y;
        long shapesize;
    };

The IDL output shows the structure of the data being published. This helps you understand what fields will appear when live data is echoed.

Step 10: Echo Live Topic Data
------------------------------

Run:

.. code-block:: bash

    echo Square

*Fast DDS Spy* will print live samples published on the ``Square`` topic.

.. code-block:: text

    >> echo Square
    ---
    {
        "color": "RED",
        "shapesize": 30,
        "x": 141,
        "y": 164
    }
    ---

    ---
    {
        "color": "RED",
        "shapesize": 30,
        "x": 135,
        "y": 159
    }
    ---

The exact values will change while *ShapesDemo* is running. Press ``Enter`` to stop printing live data and return to the *Fast DDS Spy* prompt.

> For more detailed output including topic name and type, writer GUID, partitions, and a timestamp, run:

Reviewer: ```suggestion
For more detailed output, including topic name and type, writer GUID, partitions, and a timestamp, run:
```

.. code-block:: bash

    echo Square verbose

> 

Reviewer: Add the output of the command, 2 messages from Square would be enough

*Fast DDS Spy* is now reading and displaying user data from the DDS network. This confirms that the topic data type was discovered and that live samples are being received.

.. raw:: html

    <video width=100% height=auto autoplay loop controls>
        <source src="../../../../_static/Topics.mp4">
        Your browser does not support the video tag.
    </video>
    <br></br>

Step 11: Exit Fast DDS Spy
----------------------------

Run:

.. code-block:: bash

    quit

*Fast DDS Spy* will close and return you to the normal terminal prompt.

Optional: Use a YAML Configuration File
-----------------------------------------

*Fast DDS Spy* can be configured using a YAML file. This is useful when you want to change the DDS domain, configure transports, or filter topics.

A minimal configuration example:

.. code-block:: yaml

    dds:

>       domain: 0

Reviewer: ```suggestion
      domain: 10
```

        topics                                    : Topics discovered in the network in compact format.
        topics v                                  : Topics discovered in the network.
        topics vv                                 : verbose information about Topics discovered in the network.
        topics <name>                             : Topics discovered in the network filtered by name (wildcard allowed (*)).
        topics <name> idl                         : Display the IDL type definition for topics matching <name> (wildcards allowed).
        filters                                   : Display the active filters.
        filters clear                             : Clear all the filter lists.
        filters remove                            : Remove all the filter lists.
        filter clear <category>                   : Clear <category> filter list.
        filter remove <category>                  : Remove <category> filter list.
        filter set <category> <filter_str>        : Set <category> filter list with <filter_str> as first value.
        filter add <category> <filter_str>        : Add <filter_str> in <category> filter list.

>         filter remove <category> <filter_str>     : Remove <filter_str> in <category> filter list.

Reviewer: ```suggestion
        filters clear <category>                  : Clear <category> filter list.
        filters remove <category>                 : Remove <category> filter list.
        filters set <category> <filter_str>       : Set <category> filter list with <filter_str> as first value.
        filters add <category> <filter_str>       : Add <filter_str> in <category> filter list.
        filters remove <category> <filter_str>    : Remove <filter_str> in <category> filter list.
```
